keps/sig-scheduling/5055-dra-device-taints-and-tolerations/README.md

```diff
@@ -0,0 +1,768 @@
+decision for core DRA to not require that the name is tied to one particular
+hardware instance to support hot-swapping. Admins might favor using the names
+whereas health monitoring might prefer to be specific and use a vendor-defined
+unique ID. Both are supported, which creates additional complexity.
+
+Without a kubectl extension similar to `kubectl taint nodes` the user
+experience for admins will be a bit challenging. They need to decide how to
+identify the device (by name or with a CEL expression), manually create a
+ResourceSlicePatch with a unique name, then remember to remove that
+ResourceSlicePatch again. For beta, support in `kubectl` for common
+operations may be needed.
+
+## Design Details
+
+The feature is following the approach and APIs taken for node taints and
+applies them to devices. A new controller watches tainted devices and deletes
+pods using them unless they tolerate the device taint, similar to the
+[taint-eviction-controller](https://github.com/kubernetes/kubernetes/blob/32130691a4cb8a1034b999341c40e48d197f5465/pkg/controller/tainteviction/taint_eviction.go#L81-L83). A pod which is running or has finalizers will not get removed
+immediately. Instead, the `DeletionTimestamp` gets set. That's okay for
+the purpose of this KEP:
+- The kubelet will stop any running containers and mark the pod as completed.
+- The ResourceClaim controller will remove such a completed pod from the claim's
+  `ReservedFor` and deallocate the claim once it has no consumers.
+
+Taints are cumulative as long as the key and effect pairs are different:
+- Taints defined by an admin in a ResourceSlicePatch get added to the
+  set of taints defined by the DRA driver in a ResourceSlice.
+- Taints with the same key and effect get overwritten, using the same
+  precedence as for attributes.
+
+This merging will be implemented by the same code that also
+overrides device attributes.
+
+To ensure consistency among all pods sharing a ResourceClaim, the toleration
+for taints gets added to the request in a ResourceClaim, not the pod. This also
+avoids conflicts like one pod tolerating a taint for scheduling and some other
+pod not tolerating that.
+
+Device and node taints are applied independently. A node taint applies to all
+pods on a node, whereas a device taint affects claim allocation and only those
+pods using the claim.
+
+### API
+
+The ResourceSlice content gets extended:
+
+```Go
+// BasicDevice defines one device instance.
+type BasicDevice struct {
+    ...
+
+    // If specified, the device's taints.
+    //
+    // This is an alpha field and requires enabling the DRADeviceTaints
+    // feature gate.
+    //
+    // +optional
+    // +listType=atomic
+    // +featureGate=DRADeviceTaints
+    Taints []Taint
+}
+
+// The device this Taint is attached to has the "effect" on
+// any claim and, through the claim, to pods that do not tolerate
+// the Taint.
+type Taint struct {
```

Comment:
Should we name it `DeviceTaint` to distinguish it from `v1.Taint`?

Reply:
This is the `resource.Taint` variant. It's not really specific to devices, so I prefer to not add `Device` as prefix. But I don't feel strongly about this.

Reply:
I change my mind :sweat_smile: It's used alongside DeviceAttribute and DeviceCapacity, so DeviceTaint indeed makes more sense.

Reply:
We also need to pick a maximum number of taints per device. 8 seems appropriate.